### ISSUE TITLE
anaconda-client doesn't know how to read valid metadata from non-setuptools build backends

### ISSUE THREAD

**Issue by matthewfeickert** [anaconda/anaconda-client#676]
As part of https://github.com/scientific-python/upload-nightly-action/issues/15 I'm building a test package to upload following https://packaging.python.org/en/latest/tutorials/packaging-projects/. This minimal package

```console
$ tree tests/debug_project 
tests/debug_project
├── pyproject.toml
├── README.md
└── src
    └── debug
        └── __init__.py

2 directories, 3 files
```

contains sufficient build information to be built with `hatchling` and be validated by `twine`

```console
$ python -m build . && twine check --strict dist/*
* Creating venv isolated environment...
* Installing packages in isolated environment... (hatchling)
* Getting build dependencies for sdist...
* Building sdist...
* Building wheel from sdist
* Creating venv isolated environment...
* Installing packages in isolated environment... (hatchling)
* Getting build dependencies for wheel...
* Building wheel...
Successfully built debug-0.0.1.tar.gz and debug-0.0.1-py3-none-any.whl
Checking dist/debug-0.0.1-py3-none-any.whl: PASSED
Checking dist/debug-0.0.1.tar.gz: PASSED
```

However, `anaconda-client` `v1.12.0` does not view it as having valid metadata

```
...
Processing "dist/debug-0.0.1-py3-none-any.whl"
Detecting file type...
File type is "Standard Python"
Extracting standard python attributes for upload
Error:  Trouble reading metadata from "dist/debug-0.0.1-py3-none-any.whl". Is this a valid standard python package?
```

raising (I think) the exception block here

https://github.com/Anaconda-Platform/anaconda-client/blob/bb2e957b588a07ddd806b791ac367fa88b49ccb1/binstar_client/commands/upload.py#L253-L266

As `anaconda-client` apparently doesn't agree with `twine` on what is required metadata can you please specify what the required metadata is?

**Comment by matthewfeickert** [comment]
@vshevchenko-anaconda Feel free to close given that I'm somehow unable to replicate this (in the same Docker image as yesterday(?)) from the state of https://github.com/scientific-python/upload-nightly-action/pull/27 as things are now properly failing out with

```
[ERROR] ('Authorization token is no longer valid', 401)
```

for an invalid token (as I don't want to upload for these issue discussions)

```console
$ docker run --rm -ti -v $PWD:/work:ro mambaorg/micromamba:1.4.9-bullseye-slim /bin/bash
(base) mambauser@2a53f5d1c960:/tmp$ micromamba --quiet install --channel conda-forge --yes build twine anaconda-client
(base) mambauser@2a53f5d1c960:/tmp$ python -m build --outdir ./dist /work/tests/test_package/ && twine check --strict dist/*
* Creating venv isolated environment...
* Installing packages in isolated environment... (hatchling)
* Getting dependencies for sdist...
* Building sdist...
* Building wheel from sdist
* Creating venv isolated environment...
* Installing packages in isolated environment... (hatchling)
* Getting dependencies for wheel...
* Building wheel...
Successfully built test_package-0.0.1.tar.gz and test_package-0.0.1-py3-none-any.whl
Checking dist/test_package-0.0.1-py3-none-any.whl: PASSED
Checking dist/test_package-0.0.1.tar.gz: PASSED
(base) mambauser@2a53f5d1c960:/tmp$ anaconda --token fail upload --user scientific-python-nightly-wheels ./dist/*.whl
Using Anaconda API: https://api.anaconda.org
Using "scientific-python-nightly-wheels" as upload username
Processing "./dist/test_package-0.0.1-py3-none-any.whl"
Detecting file type...
File type is "Standard Python"
Extracting standard python attributes for upload
Creating package "test-package"
[ERROR] ('Authorization token is no longer valid', 401)
(base) mambauser@2a53f5d1c960:/tmp$ 
```

it would perhaps be interesting to still know what the required metadata is for `anaconda-client`, but that's maybe a discussion for another issue.

**Comment by matthewfeickert** [comment]
@vshevchenko-anaconda Hm, apparently there is a problem though as when the GitHub Action runs with a valid token (c.f. https://github.com/scientific-python/upload-nightly-action/pull/27#discussion_r1270228235) we are getting this failure I mentioned later on in the logs.

```
...
Uploading wheels to anaconda.org...
Using Anaconda API: https://api.anaconda.org/
Using "scientific-python-nightly-wheels" as upload username
Processing "dist/test_package-0.0.1-py3-none-any.whl"
Detecting file type...
File type is "Standard Python"
Extracting standard python attributes for upload
Error:  Trouble reading metadata from "dist/test_package-0.0.1-py3-none-any.whl". Is this a valid standard python package?
Error:  Trouble reading metadata from "dist/test_package-0.0.1-py3-none-any.whl". Is this a valid standard python package?
```

**Comment by matthewfeickert** [comment]
@vshevchenko-anaconda The problem is apparently however `anaconda-client` is validating metadata is assuming it was built with `setuptools` and not any other build backend that adheres to [PEP 518](https://peps.python.org/pep-0518/).

I tested this by keeping everything else the same and switching build backends from `hatchling` to `setuptools` which then worked

```diff
--- a/tests/test_package/pyproject.toml
+++ b/tests/test_package/pyproject.toml
@@ -1,6 +1,6 @@
 [build-system]
-requires = ["hatchling"]
-build-backend = "hatchling.build"
+requires = ["setuptools>=61.0"]
+build-backend = "setuptools.build_meta"
 
 [project]
 name = "test-package"
```

This is pretty serious as this means that you're requiring a specific build backend to have build a Python package.

**Comment by matthewfeickert** [comment]
@ofek, while this isn't your problem to fix, I thought I might tag you here in the event that you can provide information that might help move this along.

**Comment by ofek** [comment]
Are you able to modify the code that checks validity so that we may see the actual traceback?

**Comment by matthewfeickert** [comment]
> Are you able to modify the code that checks validity so that we may see the actual traceback?

I would need to build and install `anaconda-client` locally to do that. That isn't impossible, but that will need to wait until next week at the earliest, unless an Anaconda engineer responds here.

**Comment by matthewfeickert** [comment]
@csoja @vshevchenko-anaconda any chance that this could get looked at in the near future to figure out why `anaconda-client` rejects metadata that is validated by `twine`? This is still happening as of `anacodna-client` `v1.12.3`.

**Comment by skupr-anaconda** [comment]
@vshevchenko-anaconda anaconda [upload] v1.12.3 fails when it tries to `rmtree()` the temporary extraction of the `info` subdirectory of the package tarball because individual files are not writable. If you make them writable the issue will be resolved.

**Comment by ifitchet** [comment]
Here's a backtrace:
```
(base) C:\Users\IanFitchet\conda-bld\win-64>anaconda --verbose --show-traceback upload msys2-perl-5.38.2-1000.tar.bz2
Using Anaconda API: https://api.anaconda.org
[DEBUG] Using conda..read_binstar_tokens
[WARNING] Token warning: expires in 9 days
Using "ifitchet" as upload username
Processing "msys2-perl-5.38.2-1000.tar.bz2"
Detecting file type...
[DEBUG] Testing if msys2-perl-5.38.2-1000.tar.bz2 is a conda package ..
[DEBUG] This is a conda package
File type is "Conda"
Extracting conda attributes for upload
[ERROR] Trouble reading metadata from "msys2-perl-5.38.2-1000.tar.bz2". Is this a valid conda package?
[ERROR]
Traceback (most recent call last):
  File "C:\Users\IanFitchet\mc1\Lib\site-packages\binstar_client\commands\upload.py", line 257, in _update_attrs
    self.__package_attrs, self.__release_attrs, self.__file_attrs = detect.get_attrs(
                                                                    ^^^^^^^^^^^^^^^^^
  File "C:\Users\IanFitchet\mc1\Lib\site-packages\binstar_client\utils\detect.py", line 286, in get_attrs
    return INSPECTORS[package_type](filename, fileobj, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\IanFitchet\mc1\Lib\site-packages\binstar_client\inspect_package\conda.py", line 162, in inspect_conda_package
    rmtree(tmpdir)
  File "C:\Users\IanFitchet\mc1\Lib\shutil.py", line 759, in rmtree
    return _rmtree_unsafe(path, onerror)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\IanFitchet\mc1\Lib\shutil.py", line 617, in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
  File "C:\Users\IanFitchet\mc1\Lib\shutil.py", line 617, in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
  File "C:\Users\IanFitchet\mc1\Lib\shutil.py", line 622, in _rmtree_unsafe
    onerror(os.unlink, fullname, sys.exc_info())
  File "C:\Users\IanFitchet\mc1\Lib\shutil.py", line 620, in _rmtree_unsafe
    os.unlink(fullname)
PermissionError: [WinError 5] Access is denied: 'C:\\Users\\IANFIT~1\\AppData\\Local\\Temp\\tmpcvt28imu\\info\\licenses\\LICENSE0.txt'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\IanFitchet\mc1\Scripts\anaconda-script.py", line 10, in <module>
    sys.exit(main())
             ^^^^^^
  File "C:\Users\IanFitchet\mc1\Lib\site-packages\binstar_client\scripts\cli.py", line 257, in main
    binstar_main(commands, args, exit_)
  File "C:\Users\IanFitchet\mc1\Lib\site-packages\binstar_client\scripts\cli.py", line 206, in binstar_main
    return arguments.main(arguments)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\IanFitchet\mc1\Lib\site-packages\binstar_client\commands\upload.py", line 61, in main
    uploader.upload(filename)
  File "C:\Users\IanFitchet\mc1\Lib\site-packages\binstar_client\commands\upload.py", line 504, in upload
    return self.upload_package(filename, package_meta)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\IanFitchet\mc1\Lib\site-packages\binstar_client\commands\upload.py", line 516, in upload_package
    meta._update_attrs(parser_args=self.arguments)  # pylint: disable=protected-access
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\IanFitchet\mc1\Lib\site-packages\binstar_client\commands\upload.py", line 266, in _update_attrs
    raise errors.BinstarError(message) from error
binstar_client.errors.BinstarError: Trouble reading metadata from "msys2-perl-5.38.2-1000.tar.bz2". Is this a valid conda package?
```

**Comment by mattkram** [comment]
Regarding the last two comments about permissions, that should be addressed by #724.

@matthewfeickert apologies for the delay. Are you still having your original issue? I created a simple package based on your example above and was able to upload a wheel to anaconda.org when the backend was both `setuptools` and `hatch`.

**Comment by matthewfeickert** [comment]
> Are you still having your original issue? I created a simple package based on your example above and was able to upload a wheel to anaconda.org when the backend was both `setuptools` and `hatch`.

No. I'm not sure what happened, but even with `anaconda-client==1.12.0` I'm now able to run the same workflow with the same code and it will now validate and upload. So that's very confusing. :face_with_diagonal_mouth: 

As this is no longer reproducible I'm going to close this.
